builtin/method_str.py

```diff
@@ -130,3 +135,148 @@ def Call(self, rd):
+
+
+class Replace(vm._Callable):
+
+    def __init__(self, mem, expr_ev):
+        # type: (state.Mem, expr_eval.ExprEvaluator) -> None
+        self.mem = mem
+        self.expr_ev = expr_ev
+
+    def EvalSubstExpr(self, expr, blame_loc):
+        # type: (value.Expr, loc_t) -> str
+        res = self.expr_ev.EvalExpr(expr.e, blame_loc)
+        if res.tag() == value_e.Str:
+            return cast(value.Str, res).s
+
+        raise error.TypeErr(res, "expected expr to eval to a string",
+                            blame_loc)
+
+    def Call(self, rd):
+        # type: (typed_args.Reader) -> value_t
+        """
+        s => replace(string_val, subst_str, count=-1)
+        s => replace(string_val, subst_expr, count=-1)
+        s => replace(eggex_val, subst_str, count=-1)
+        s => replace(eggex_val, subst_expr, count=-1)
+        """
+        string = rd.PosStr()
+
+        string_val = None  # type: value.Str
+        eggex_val = None  # type: value.Eggex
+        subst_str = None  # type: value.Str
+        subst_expr = None  # type: value.Expr
+
+        pattern = rd.PosValue()
+        with tagswitch(pattern) as case:
+            if case(value_e.Eggex):
+                # HACK: mycpp will otherwise generate:
+                #  value::Eggex* eggex_val ...
+                eggex_val_ = cast(value.Eggex, pattern)
+                eggex_val = eggex_val_
+
+            elif case(value_e.Str):
+                string_val_ = cast(value.Str, pattern)
+                string_val = string_val_
+
+            else:
+                raise error.TypeErr(pattern,
+                                    'expected pattern to be Eggex or Str',
+                                    rd.LeftParenToken())
+
+        subst = rd.PosValue()
+        with tagswitch(subst) as case:
+            if case(value_e.Str):
+                subst_str_ = cast(value.Str, subst)
+                subst_str = subst_str_
+
+            elif case(value_e.Expr):
+                subst_expr_ = cast(value.Expr, subst)
+                subst_expr = subst_expr_
+
+            else:
+                raise error.TypeErr(subst,
+                                    'expected substitution to be Str or Expr',
+                                    rd.LeftParenToken())
+
+        count = rd.NamedInt("count", -1)
+        rd.Done()
+
+        if count == 0:
+            return value.Str(string)
+
+        if string_val:
+            if subst_str:
+                s = subst_str.s
+            if subst_expr:
+                s = self.EvalSubstExpr(subst_expr, rd.LeftParenToken())
+
+            result = string.replace(string_val.s, s, count)
+
+            return value.Str(result)
+
+        if eggex_val:
+            ere = regex_translate.AsPosixEre(eggex_val)
+            cflags = regex_translate.LibcFlags(eggex_val.canonical_flags)
+
+            pos = 0
+            parts = []  # type: List[str]
+            replace_count = 0
+            while pos < len(string):
```

Comment:
Some comments about the algorithm here would help

I'm wondering if we can consolidate it with anything in `func_eggex`

Though we don't need to do that to start, since the tests look good here